File: app/src/main/java/org/tvheadend/tvhclient/ui/features/dvr/ChannelListSelectionAdapter.java
package org.tvheadend.tvhclient.ui.features.dvr;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.domain.entity.Channel;
import org.tvheadend.tvhclient.util.MiscUtils;

import java.lang.ref.WeakReference;
import java.util.List;

// TODO convert to data binding

public class ChannelListSelectionAdapter extends RecyclerView.Adapter<ChannelListSelectionAdapter.ViewHolder> {

    private final WeakReference<Context> context;
    private Callback callback;
    private final List<Channel> channelList;

    interface Callback {
        void onItemClicked(Channel channel);
    }

    ChannelListSelectionAdapter(Context context, List<Channel> channelList) {
        this.context = new WeakReference<>(context);
        this.channelList = channelList;
    }

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.channel_list_selection_dialog_adapter, parent, false);
        return new ViewHolder(view, this);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        final Channel channel = channelList.get(position);
        if (channel != null) {
            holder.itemView.setTag(channel);
            Context context = this.context.get();
            if (context != null && holder.iconImageView != null && !TextUtils.isEmpty(channel.getIcon())) {
                Picasso.get()
                        .load(MiscUtils.getIconUrl(context, channel.getIcon()))
                        .into(holder.iconImageView);
            }
            if (holder.titleTextView != null) {
                holder.titleTextView.setText(channel.getName());
                holder.titleTextView.setTag(position);
            }
        }
    }

    @Override
    public int getItemCount() {
        return channelList.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        final ImageView iconImageView;
        final TextView titleTextView;
        final ChannelListSelectionAdapter channelListAdapter;

        ViewHolder(View view, ChannelListSelectionAdapter adapter) {
            super(view);
            iconImageView = view.findViewById(R.id.icon);
            titleTextView = view.findViewById(R.id.title);
            channelListAdapter = adapter;
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (channelListAdapter != null && channelListAdapter.callback != null) {
                Channel channel = channelListAdapter.channelList.get(getAdapterPosition());
                channelListAdapter.callback.onItemClicked(channel);
            }
        }
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/dvr/ChannelListSelectionAdapter.kt
package org.tvheadend.tvhclient.ui.features.dvr

import android.content.Context
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.domain.entity.Channel
import org.tvheadend.tvhclient.util.getIconUrl
import java.lang.ref.WeakReference

// TODO convert to data binding

class ChannelListSelectionAdapter internal constructor(context: Context, private val channelList: List<Channel>) : RecyclerView.Adapter<ChannelListSelectionAdapter.ViewHolder>() {

    private val context: WeakReference<Context> = WeakReference(context)
    private var callback: Callback? = null

    interface Callback {
        fun onItemClicked(channel: Channel)
    }

    fun setCallback(callback: Callback) {
        this.callback = callback
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.channel_list_selection_dialog_adapter, parent, false)
        return ViewHolder(view, this)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val channel = channelList[position]
        holder.itemView.tag = channel
        val context = this.context.get()
        if (context != null && holder.iconImageView != null && !TextUtils.isEmpty(channel.icon)) {
            Glide.with(context)
                    .load(getIconUrl(context, channel.icon))
                    .into(holder.iconImageView)
        }
        if (holder.titleTextView != null) {
            holder.titleTextView.text = channel.name
            holder.titleTextView.tag = position
        }
    }

    override fun getItemCount(): Int {
        return channelList.size
    }

    class ViewHolder internal constructor(view: View, private val channelListAdapter: ChannelListSelectionAdapter?) : RecyclerView.ViewHolder(view), View.OnClickListener {
        internal val iconImageView: ImageView? = view.findViewById(R.id.icon)
        internal val titleTextView: TextView? = view.findViewById(R.id.title)

        init {
            view.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            if (channelListAdapter?.callback != null) {
                val channel = channelListAdapter.channelList[adapterPosition]
                channelListAdapter.callback!!.onItemClicked(channel)
            }
        }
    }
}
